drivers/sensor/bma4xx/bma4xx_emul.c

```diff
@@ -0,0 +1,428 @@
+			 size_t count)
+{
+	struct bma4xx_emul_data *data = target->data;
+
+	__ASSERT_NO_MSG(reg_addr + count <= NUM_REGS);
+	memcpy(data->regs + reg_addr, val, count);
+}
+
+void bma4xx_emul_get_reg(const struct emul *target, uint8_t reg_addr, uint8_t *val, size_t count)
+{
+	struct bma4xx_emul_data *data = target->data;
+
+	__ASSERT_NO_MSG(reg_addr + count <= NUM_REGS);
+	memcpy(val, data->regs + reg_addr, count);
+}
+
+bool bma4xx_emul_is_accel_enabled(const struct emul *emul)
+{
+	struct bma4xx_emul_data *data = emul->data;
+
+	return data->accel_enabled;
+}
+
+uint8_t bma4xx_emul_get_accel_range(const struct emul *emul)
+{
+	struct bma4xx_emul_data *data = emul->data;
+
+	return data->accel_range;
+}
+
+uint32_t bma4xx_emul_get_odr(const struct emul *emul)
+{
+	struct bma4xx_emul_data *data = emul->data;
+
+	/*
+	 * This function deliberately differs from bma4xx_reg_to_odr() to provide
+	 * an obviously-correct reference for tests.
+	 */
+	switch (data->odr_raw) {
+	case 1:
+		return 781; /* 25/32 Hz */
+	case 2:
+		return 1562; /* 25/16 Hz */
+	case 3:
+		return 3125;
+	case 4:
+		return 6250;
+	case 5:
+		return 12500;
+	case 6:
+		return 25000;
+	case 7:
+		return 50000;
+	case 8:
+		return 100000;
+	case 9:
+		return 200000;
+	case 10:
+		return 400000;
+	case 11:
+		return 800000;
+	case 12:
+		return 1600000;
+	default:
+		LOG_ERR("ODR register value %#x is reserved", data->odr_raw);
+		return 0;
+	}
+}
+
+void bma4xx_emul_set_accel_data(const struct emul *emul, int x, int y, int z)
+{
+	struct bma4xx_emul_data *data = emul->data;
+
+	data->acceleration[0] = x;
+	data->acceleration[1] = y;
+	data->acceleration[2] = z;
+}
+
+uint8_t bma4xx_emul_get_nv_conf(const struct emul *emul)
+{
+	struct bma4xx_emul_data *data = emul->data;
+
+	return data->nv_config;
+}
+
+bool bma4xx_emul_is_fifo_enabled(const struct emul *emul)
+{
+	struct bma4xx_emul_data *data = emul->data;
+
+	return data->fifo_enabled;
+}
+
+void bma4xx_emul_set_fifo_data(const struct emul *emul, const uint8_t *fifo_data, uint16_t data_sz)
+{
+	struct bma4xx_emul_data *data = emul->data;
+
+	__ASSERT(data_sz % 6 == 0, "FIFO data should be an integer number of frames");
+	data->fifo_data = fifo_data;
+	data->fifo_available = data_sz;
+}
+
+uint8_t bma4xx_emul_get_interrupt_config(const struct emul *emul, uint8_t *int1_io_ctrl,
+					 bool *latched_mode)
+{
+	struct bma4xx_emul_data *data = emul->data;
+
+	*int1_io_ctrl = data->int1_io_ctrl;
+	*latched_mode = data->interrupt_mode_latched;
+	return data->int_map_data;
+}
+
+static int bma4xx_emul_read_byte(const struct emul *target, int reg, uint8_t *val, int bytes)
+{
+	struct bma4xx_emul_data *data = target->data;
+
+	if (reg != BMA4XX_REG_FIFO_DATA) {
+		/*
+		 * Burst reads autoincrement register addresses, except for
+		 * FIFO data which reads from the FIFO instead.
+		 */
+		reg += bytes;
+	}
+
+	switch (reg) {
+	case BMA4XX_REG_CHIP_ID:
+		*val = BMA4XX_CHIP_ID; /* BMA4XX_CHIP_ID */
```

Comment:
Lets store the chip ID in the `regs[]` array at init. Some tests might want to mess with this to check what happens if the wrong chip is placed on the board.